pkg/data/generic.go

```diff
@@ -203,31 +196,81 @@ func (s *GenericSync) sync(resource dynamic.NamespaceableResourceInterface, quit
 }
 
 func (s *GenericSync) syncAdd(obj runtime.Object) error {
-	m, err := meta.Accessor(obj)
+	path, err := s.objPath(obj)
 	if err != nil {
 		return err
 	}
-	name := m.GetName()
-	var path = m.GetName()
-	if s.ns.Namespaced {
-		path = m.GetNamespace() + "/" + name
-	}
 	return s.opa.PutData(path, obj)
 }
 
 func (s *GenericSync) syncRemove(obj runtime.Object) error {
-	m, err := meta.Accessor(obj)
+	path, err := s.objPath(obj)
+	if err != nil {
+		return err
+	}
+	return s.opa.PatchData(path, opa_client.DataPatchOpRemove, nil)
+}
+
+func (s *GenericSync) syncAll(objs []unstructured.Unstructured) error {
+
+	// Build a list of patches to apply.
+	var patches []opa_client.DataPatch
+
+	createdSubPaths := map[string]struct{}{}
+
+	for _, obj := range objs {
+
+		path, err := s.objPath(&obj)
+		if err != nil {
+			return err
+		}
+
+		var subPath string
+		for _, segment := range strings.Split(path, "/") {
+			subPath = filepath.Join(subPath, segment)
+
+			// Skip ones we already created a patch for
+			if _, alreadyExists := createdSubPaths[subPath]; alreadyExists {
+				continue
+			}
+
+			// Add an empty interface at the sub-path. JSON patch "add" requires that
+			// the document we are adding to exists, in this case it is the object at
+			// each sub-path up to and including the full path the value is added to.
+			var empty interface{}
+			empty = map[string]interface{}{}
+			patches = append(patches, s.opa.NewPatch(subPath, opa_client.DataPatchOpAdd, &empty))
```

Comment:
Is this going to do an extra add for every single object? E.g., the resource type and namespace collections will initialize once, but for each object, we'll an empty value here and the the final value below? Can we just skip the last segment?

Reply:
Yep, I'll adjust to skip the last segment. Unit tests (coming soon ™️ ) to validate the patch generation should catch this type of thing too.

Reply:
Fixed in latest update.